x-pack/plugins/alerting/server/rules_client/rules_client.ts

```diff
@@ -1100,6 +1176,224 @@ export class RulesClient {
+    const errors: BulkEditError[] = [];
+    const apiKeysToInvalidate: string[] = [];
+
+    for await (const response of rulesFinder.find()) {
+      await pMap(
+        response.saved_objects,
+        async (rule) => {
+          try {
+            if (rule.attributes.apiKey) {
+              apiKeysToInvalidate.push(rule.attributes.apiKey);
+            }
+
+            const ruleType = this.ruleTypeRegistry.get(rule.attributes.alertTypeId);
+
+            let attributes = rule.attributes;
+            let ruleActions = {
+              actions: this.injectReferencesIntoActions(
+                rule.id,
+                rule.attributes.actions,
+                rule.references || []
+              ),
+            };
+            for (const action of actions) {
+              if (action.field === 'actions') {
+                await this.validateActions(ruleType, action.value);
+                ruleActions = this.applyBulkEditAction(action, ruleActions);
+              }
+
+              attributes = this.applyBulkEditAction(action, attributes);
+            }
+
+            const ruleParams = paramsModifier
+              ? paramsModifier(rule.attributes.params as Params)
+              : rule.attributes.params;
+
+            // validate rule params
+            const validatedAlertTypeParams = validateRuleTypeParams(
+              ruleParams,
+              ruleType.validate?.params
+            );
+            validateMutatedRuleTypeParams(validatedAlertTypeParams, ruleType.validate?.params);
+
+            // validate schedule interval
+            if (attributes.schedule.interval) {
+              const intervalInMs = parseDuration(attributes.schedule.interval as string);
+              if (intervalInMs < this.minimumScheduleIntervalInMs) {
+                throw Boom.badRequest(
+                  `Error updating rule: the interval is less than the allowed minimum interval of ${this.minimumScheduleInterval}`
+                );
+              }
+            }
+
+            const {
+              actions: rawAlertActions,
+              references,
+              params: updatedParams,
+            } = await this.extractReferences(
+              ruleType,
+              ruleActions.actions,
+              validatedAlertTypeParams
+            );
+
+            // create API key
+            const username = await this.getUserName();
+            let createdAPIKey = null;
+            try {
+              createdAPIKey = attributes.enabled
+                ? await this.createAPIKey(this.generateAPIKeyName(ruleType.id, attributes.name))
+                : null;
+            } catch (error) {
+              throw Error(`Error updating rule: could not create API key - ${error.message}`);
+            }
+            const apiKeyAttributes = this.apiKeyAsAlertAttributes(createdAPIKey, username);
+
+            // get notifyWhen
+            const notifyWhen = getAlertNotifyWhenType(
```

Comment:
I can see that attributes.notifyWhen can be `null` and this function effectively eliminates null from it.
notifyWhen comes from rules attributes received from results of createPointInTimeFinderAsInternalUser
```typescript
    for await (const response of rulesFinder.find()) {
      await pMap(
        response.saved_objects,
        async (rule) => {
``

How we would be able to make sure it's not null and not calling `getAlertNotifyWhenType`?

Reply:
Ah, this might be a problem with our TypeScript definitions. The existing saved-objects should all have `notifyWhen` set. In 7.11 we put a migration to set a value to all existing documents [ref](https://github.com/elastic/kibana/blob/45c4e7dac1390df210d211fab57a27415be7d93e/x-pack/plugins/alerting/server/saved_objects/migrations.ts#L95) and then set `notifyWhen` via the create API for backwards compatibility when it's not provided.

Reply:
So, we can remove `null` here https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/types.ts#L247

Or leave this `getAlertNotifyWhenType`  as it is?

Reply:
After taking a look, let's leave it as is 👍